gax/src/main/java/com/google/api/gax/batching/BlockingSemaphore.java

```diff
@@ -43,31 +46,100 @@ private static void checkNotNegative(long l) {
   BlockingSemaphore(long permits) {
     checkNotNegative(permits);
     this.currentPermits = permits;
+    this.limit = permits;
   }
 
-  public synchronized void release(long permits) {
+  public void release(long permits) {
     checkNotNegative(permits);
-
-    currentPermits += permits;
-    notifyAll();
+    synchronized (updateLock) {
+      // If more permits are returned than what was originally set, we need to add these extra
+      // permits to the limit too
+      currentPermits += permits;
+      if (currentPermits > limit) {
+        limit = currentPermits;
+      }
+    }
+    synchronized (waitLock) {
+      waitLock.notifyAll();
+    }
   }
 
-  public synchronized boolean acquire(long permits) {
+  public boolean acquire(long permits) {
     checkNotNegative(permits);
 
     boolean interrupted = false;
-    while (currentPermits < permits) {
-      try {
-        wait();
-      } catch (InterruptedException e) {
-        interrupted = true;
+    while (!interrupted) {
```

Comment:
(Partially this comment applies even to the original implementation, pushed in part 1, which I LGTM'ed but I missed it during reiew of part 1)

It looks like the only way the thread, which entered this loop can exit is is by receiving interrupt signal. It looks very dangerous not only for correctness of this new feature but for correctness of the rest of gax in an indirect way.

In general, pretty much all the code in gax-java is executed in the context of the thread executor pool, which has 4 threads by default (or number of current CPUs if > 4). My biggest concern is that once code enters this block, it means **one of those 4 threads is completely owned by this block** and there are only two possible ways out:
- this code will never end and will keep attempting to wait forever;
- this thread gets interrupted, then the execution reaches line 91 and `Thread.currentTrhead.interrupt()` is called.

Then it is not clear what will happen with this thread after `interrupt()` is called, as it only sets the interrupt flag to `true`, and does not guarantee that the thread does not get "reinterrupted" (interrupt flag cleared). 

Also, if the thread actually gets terminated, it means it will have to be re-created in the executor pool, basically defeating the purpose of using thread pool (compared to simply spawning new thread for every shingle task). In other words this code block messes with executors pool in a very brutal way: 
- 4 parallel entries in this block and the executor pool is dead (assuming 4 is the number of configured threads)
- if the execution flow assumes frequent entering/leaving this code block, it would result into frequent thread_interrupt->thread_kill->thread_create, which is a heavy operation and is not how threads arfe supposed to be used in the context of a thread pool (threads must be shared by tasks, and tasks should never dictate the lifecycle of a thread it is executed in).

Basically this code block does not "respect" the other potential tasks in the pool, assuming that the thread it executes in now belongs to it and will never be released back to the pool.

Reply:
@igorbernstein2 PTAL.

Reply:
Reverted back to the original logic, which will wait for the permits and interrupt. And it has always been implemented this way  https://github.com/googleapis/gax-java/blame/master/gax/src/main/java/com/google/api/gax/batching/BlockingSemaphore.java#L59-L70

>  4 parallel entries in this block and the executor pool is dead 
This class is used in FlowController, so wherever it's called, it'll want to block the thread from doing anything to throttle?

> if the execution flow assumes frequent entering/leaving this code block, it would result into frequent thread_interrupt->thread_kill->thread_create, which is a heavy operation and is not how threads are supposed to be used in the context of a thread pool
Sorry again for the confusing changes from using 2 locks (which I've reverted back). But in both implementations, the expected way of leaving this code block is when there are more permits then what's asked by the caller, and interrupt should rarely happen.